desc/objectives/objective_funs.py

```diff
@@ -740,14 +663,55 @@ def weights(self):
             [jnp.ones(obj.dim_f) * obj.weight for obj in self.objectives]
         )
 
+    @property
+    def _all_things(self):
+        """list: all things known to this objective, used and unused."""
+        if not hasattr(self, "_extra_things"):
+            self._extra_things = []
+        return [obj.things for obj in self.objectives] + self._extra_things
+
+    @property
+    def things(self):
+        """list: Optimizable things that this objective is tied to."""
+        errorif(
+            not hasattr(self, "_flatten"),
+            RuntimeError,
+            "ObjectiveFunction must be built with ObjectiveFunction.build() first",
+        )
+        return self._flatten(self._all_things)
+
+    @things.setter
+    def things(self, new):
+        if not isinstance(new, (tuple, list)):
+            new = [new]
+        assert all(isinstance(x, Optimizable) for x in new)
+        # in general this is a hard problem, since we don't really know which object
+        # to replace with which if there are multiple of the same type, but we can
+        # do our best and throw an error if we can't figure it out here.
+        expected_types = [type(a) for a in self.things]
+        got_types = [type(a) for a in new]
+        errorif(
+            expected_types != got_types,
+            TypeError,
+            "Cannot unambiguously parse Optimizable objects to individual Objectives, "
+            + f"expected types {expected_types} but got types {got_types}. "
+            + "Try setting Objective.things on each sub Objective individually then "
+            + "call ObjectiveFunction.build().",
+        )
+        # now we know that new and self.things contains instances of unique classes, so
+        # we should be able to just replace like with like
+        things = self._unflatten(new)
+        for obj, t in zip(self.objectives, things[: -len(self._extra_things)]):
+            obj.things = t
```

Comment:
if things is changed, if any of them have differing resolutions from the original things, the objectivefunction must  be rebuilt. I think the setter should set the `self._built` flag to `False` if things results in a change (or even just whenever things is called) to enforce this and try to pre-empt dimension mismatches due to things changing resolution but the objective not being rebuilt

Reply:
sounds good